### PR TITLE
correlate_partial()

### PR DESCRIPTION
**Pull request by LKobilke**
Implements the `correlate_partial` function based on `ppcor::pcor.test`.

Accepts exactly three numeric variables and computes pairwise partial correlations (pearson, kendall, spearman) for all combinations, i.e. pairs, of specified variables while controlling for the third variable. Stops and provides a case-specific warning if too little or too many variables or non-numeric variables are provided.
If input data is grouped, groups are dropped and a warning is issued.

Combinations of pairs are calculated using `combinat::permn`.

Returns a tibble.

Currently missing is the additional model as an attribute. This is added by @joon-e, see #24 .

### CODE REVIEW

**Comment by MarHai** [comment]
Merge should aim at [`devel`](https://github.com/joon-e/tidycomm/tree/devel), not `master`.

**Comment by LKobilke** [comment]
> Code looks good and validates well.
> 
> However, I was expecting an extension to the `correlate` function with a parameter `partial = TRUE`. Why did you change this plan?

Good point. I've just pushed a new commit that implements a change to the `correlate()` function, i.e., adds the parameter `partial = TRUE`, and makes `correlate_partial()` an internal function. I was initially hesitant to do this since `correlate()` will grab all numeric variables if none are supplied. My concern was that this could lead to lots of information overload when reporting all possible x,y,z combinations. To avoid information overload `correlate(partial = FALSE)` will now work with two, three, or more variables, but `correlate(partial = TRUE)` will only accept three variables and throw a warning otherwise. This difference in behavior may seem inconsistent and I felt that a clear distinction between bivariate correlation and partial correlation might help to reduce this inconsistency.

Despite my initial concerns, I see the advantage to expand on the `correlate()` function. I'm open to keeping this version.

File: tests/testthat/test-correlate_partial.R [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  t <- WoJ %>% correlate_partial(ethics_1, ethics_2, ethics_3)
+
+  expect_true(tibble::is_tibble(t))
+  expect_equal(dim(t), c(3, 6))
+})
+
+test_that("correlate_partial() returns tibble for other methods", {
+
+  t1 <- WoJ %>% correlate_partial(ethics_1, ethics_2, ethics_3, method = "spearman")
+  t2 <- WoJ %>% correlate_partial(ethics_1, ethics_2, ethics_3, method = "kendall")
```
Comment (MarHai) [inline tests/testthat/test-correlate_partial.R:16]: Lines 15 and 16 require updates due to the latest changes.

File: tests/testthat/test-correlate_partial.R [hunk 2]
```diff
@@ -0,0 +1,61 @@
+context("correlate_partial")
+
+# Main functions
+
+test_that("correlate_partial() returns tibble", {
+
+  t <- WoJ %>% correlate_partial(ethics_1, ethics_2, ethics_3)
```
Comment (MarHai) [inline tests/testthat/test-correlate_partial.R:7]: Requires updates due to the latest changes.

File: tests/testthat/test-correlate_partial.R [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  expect_true("rho" %in% names(t1))
+  expect_equal(dim(t1), c(3, 6))
+
+  expect_true(tibble::is_tibble(t2))
+  expect_true("tau" %in% names(t2))
+  expect_equal(dim(t2), c(3, 6))
+})
+
+test_that("correlate_partial() calculates correct results for Pearson correlations", {
+  result <- WoJ %>% correlate_partial(ethics_1, ethics_2, ethics_3)
```
Comment (MarHai) [inline tests/testthat/test-correlate_partial.R:28]: Requires updates due to the latest changes.

File: tests/testthat/test-correlate_partial.R [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  partial_cor_numerator <- cor(WoJ$ethics_1, WoJ$ethics_2, method = "pearson") -
+                 cor(WoJ$ethics_1, WoJ$ethics_3, method = "pearson") *
+                 cor(WoJ$ethics_2, WoJ$ethics_3, method = "pearson")
+  partial_cor_denumerator <- sqrt((1-(cor(WoJ$ethics_1, WoJ$ethics_3, method = "pearson")^2))*(1-(cor(WoJ$ethics_3, WoJ$ethics_2, method = "pearson")^2)))
+  partial_cor <- partial_cor_numerator / partial_cor_denumerator
+  expect_equal(round(result$r[1], digits = 3), round(partial_cor, digits = 3))
+})
+
+test_that("correlate_partial() works with tidyselect helpers", {
+  expect_error(WoJ %>% correlate_partial(tidyselect::starts_with("ethics")),
```
Comment (MarHai) [inline tests/testthat/test-correlate_partial.R:38]: Requires updates due to the latest changes.

File: tests/testthat/test-correlate_partial.R [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+               "The computation cannot be performed due to an excessive number of variables provided. Please provide exactly three variables.")
+})
+
+## Possible errors
+
+test_that("Error is thrown when number of provided variables is less than three", {
+  expect_error(WoJ %>% correlate_partial(ethics_1, ethics_2),
```
Comment (MarHai) [inline tests/testthat/test-correlate_partial.R:45]: Requires updates due to the latest changes.

File: tests/testthat/test-correlate_partial.R [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+               "The computation cannot be performed because there are not enough variables provided. Please provide exactly three variables.")
+})
+
+test_that("Error is thrown when number of provided variables is more than three", {
+  expect_error(WoJ %>% correlate_partial(ethics_1, ethics_2, ethics_3, ethics_4),
```
Comment (MarHai) [inline tests/testthat/test-correlate_partial.R:50]: Requires updates due to the latest changes.

File: tests/testthat/test-correlate_partial.R [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+               "The computation cannot be performed due to an excessive number of variables provided. Please provide exactly three variables.")
+})
+
+test_that("Error is thrown when method is not one of 'pearson', 'kendall' or 'spearman'", {
+  expect_error(WoJ %>% correlate_partial(ethics_1, ethics_2, ethics_3, method = "wrong"),
+               "Method must be one of \"pearson\", \"kendall\" or \"spearman\".")
+})
+
+test_that("Error is thrown when non-numeric variables are passed", {
+  expect_error(fbposts %>% correlate_partial(type, n_pictures,pop_elite))
```
Comment (MarHai) [inline tests/testthat/test-correlate_partial.R:60]: Requires updates due to the latest changes.

File: R/correlation.R [hunk 8]
```diff
@@ -17,9 +21,20 @@
 #' @examples
 #' WoJ %>% correlate(ethics_1, ethics_2, ethics_3)
 #' WoJ %>% correlate()
+#' WoJ %>% correlate(ethics_1, ethics_2, ethics_3, partial = TRUE)
```
Comment (MarHai) [inline R/correlation.R:24]: Maybe a slightly more realistic examlpe would be helpful. For example:

`WoJ %>% correlate(autonomy_selection, autonomy_emphasis, work_experience, partial = TRUE)`

**Pull request merged**